### ISSUE TITLE
AssemblyLoadContext.Unload causes System.ExecutionEngineException during debug in VS

### ISSUE THREAD

**Issue by ststeiger** [dotnet/runtime#37053]
.NET Core 3.1
Windows 10 (DE)
Visual Studio 2019 Community (2) 16.5.0

I have the VB class posted below. 
It executes dynamically compiled code loaded with 
```
System.Runtime.Loader.AssemblyLoadContext context = new Parameters.CollectibleAssemblyLoadContext();
System.IO.Stream ms = new System.IO.MemoryStream(result);
System.Reflection.Assembly assembly = context.LoadFromStream(ms);
```

It works fine (or at least so it seems). 
But when I execute ExpressionEvaluator.Test() in step-through mode, I get 

> System.ExecutionEngineException. <br />
> Fatal Error. Internal CLR error. (0x80131506)

If I set a breakpoint further down, then it works (until the next dispose). 
If I just run it with dotnet run, then it works. 

The problem seems to be the dispose at `} // evaluator1`.
In the called AbstractEvaluator.Dispose, there is this code 

```
If Me.LoadContext IsNot Nothing Then
    Me.LoadContext.Unload()
    Me.LoadContext = Nothing
End If
```
which unloads the AssemblyLoadContext. 
My guess is the problem is caused during this. 
Maybe it's actually a bug in VisualStudio or in ExecutionEngine - I don't know. 

(See screenshots at bottom)
Also, Visual Studio says "Application in Hold mode" (or whatever the actual translation is). 
Your app is on hold because at present there is no code executed which is supported by the Debug-Engine. (actual English translation may differ)
Then it has a link/button "Continue execution", but that does nothing. 

```
namespace ReportTester
{


    public static partial class ExpressionEvaluator
    {


        public static void Test()
        {
            DirectEvaluation();


            object[] objs = new string[] { };
            string joined = Join(objs, ",");
            System.Console.WriteLine(joined);


            using (Parameters.AbstractEvaluator evaluator1 = CreateEvaluator("1+1", ""))
            {
                object obj1 = evaluator1.EvaluateExpression();
                System.Console.WriteLine(obj1);
            } // evaluator1


            using (Parameters.AbstractEvaluator evaluator2 = CreateEvaluator("Parameters!in_standort.Value", ""))
            {
                evaluator2.Parameters["in_standort"] = new ReportTester.Parameters.ReportParameter(new ReportTester.Parameters.ParameterValue() { Label = "abc", Value = "456" });

                evaluator2.Parameters["in_standort"] = new ReportTester.Parameters.ReportParameter(new ReportTester.Parameters.ParameterValue()
                {
                    Label = "abc",
                    Value = "456"
                }, new ReportTester.Parameters.ParameterValue()
                {
                    Label = "def",
                    Value = "123"
                });

                object obj2 = evaluator2.EvaluateExpression();
                System.Console.WriteLine(obj2);
                System.Console.WriteLine(Join(obj2, ","));
                string str = System.Convert.ToString(obj2, System.Globalization.CultureInfo.InvariantCulture);
                System.Console.WriteLine(str);
            } // evaluator2


            System.Console.WriteLine(System.Environment.NewLine);
            System.Console.WriteLine(" --- Press any key to continue --- ");
            System.Console.ReadKey();
        }


        public static void DirectEvaluation()
        {
            object obj1 = Evaluate("1+1", "");
            System.Console.WriteLine(obj1);

            object obj2 = Evaluate("Parameters!in_standort.Value", "",
                delegate (ReportTester.Parameters.AbstractEvaluator evaluator2)
                {
                    evaluator2.Parameters["in_standort"] = new ReportTester.Parameters.ReportParameter(new ReportTester.Parameters.ParameterValue() { Label = "abc", Value = "456" }
                    );

                    evaluator2.Parameters["in_standort"] = new ReportTester.Parameters.ReportParameter(new ReportTester.Parameters.ParameterValue()
                    {
                        Label = "abc",
                        Value = "456"
                    }, new ReportTester.Parameters.ParameterValue()
                    {
                        Label = "def",
                        Value = "123"
                    });
                }
            );

            System.Console.WriteLine(Join(obj2, ","));
            System.Console.WriteLine(obj2);
        }


#if NETCOREAPP

        public static System.AppDomain CreateAppDomain()
        {
            System.AppDomain temporaryAppDomain = null;
            return temporaryAppDomain;
        }

#else
        public static System.AppDomain CreateAppDomain()
        {
            string domainName = System.Guid.NewGuid().ToString();
            System.AppDomain temporaryAppDomain = System.AppDomain.CreateDomain(domainName, System.AppDomain.CurrentDomain.Evidence, 
                new System.AppDomainSetup()
                {
                    ApplicationBase = System.AppDomain.CurrentDomain.BaseDirectory
                }
                ,new System.Security.PermissionSet(System.Security.Permissions.PermissionState.Unrestricted)
            );

            return temporaryAppDomain;
        }

#endif


        public static string Join(object obj, string sep)
        {
            return Join((object[])obj, sep);
        }


        public static string Join(object[] obj, string sep)
        {
            string ret = null;

            if (obj == null || obj.Length == 0)
                return ret;

            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            for (int i = 0; i < obj.Length; ++i)
            {
                sb.Append(System.Convert.ToString(obj[i], System.Globalization.CultureInfo.InvariantCulture));
                if (obj.Length - 1 != i)
                    sb.Append(sep);
            }

            ret = sb.ToString();
            sb.Length = 0;
            sb = null;

            return ret;
        }

        
    } // End Class ExpressionEvaluator


} // End Namespace 
```

```
Namespace ReportTester.Parameters



    Public Delegate Sub Evaluate_Callback_t(e As AbstractEvaluator)


    Public MustInherit Class AbstractEvaluator
        Inherits System.MarshalByRefObject
        Implements System.IDisposable


        Public Domain As System.AppDomain
        Public LoadContext As System.Runtime.Loader.AssemblyLoadContext
        Public Stream As System.IO.Stream


        Public Parameters As ParameterCollection
        Private disposedValue As Boolean ' Dient zur Erkennung redundanter Aufrufe.


        Protected Overridable Sub Dispose(disposing As Boolean)
            If Not disposedValue Then
                If disposing Then
                    ' TODO: verwalteten Zustand (verwaltete Objekte) entsorgen.

                    If Me.LoadContext IsNot Nothing Then
                        Me.LoadContext.Unload()
                        Me.LoadContext = Nothing
                    End If

                    If Me.Stream IsNot Nothing Then
                        Me.Stream.Dispose()
                        Me.Stream = Nothing
                    End If

                    'If Parameters.m_parameterValues IsNot Nothing Then
                    '    Parameters.m_parameterValues.Clear()
                    '    Parameters.m_parameterValues = Nothing
                    'End If

                    ' This is for backwards compatibility with .NET Framwork 2.0 and CodeDom
                    '  Does not get called in .NET Core which uses LoadContext instead of AppDomain
                    If Me.Domain IsNot Nothing Then
                        ' https://stackoverflow.com/questions/7793074/unload-an-appdomain-while-using-idisposable
                        Dim thread As System.Threading.Thread = New System.Threading.Thread(
                          Sub(obj As System.AppDomain)
                              Try
                                  ' System.Threading.Thread.Sleep(1000)
                                  System.AppDomain.Unload(obj)
                              Catch ex As System.Threading.ThreadAbortException
                                  ' System.Console.WriteLine(ex.Message)
                                  System.GC.Collect()
                                  System.GC.WaitForPendingFinalizers()
                              End Try
                          End Sub
                        )
                        thread.IsBackground = True
                        thread.Start(Me.Domain)
                        Me.Domain = Nothing
                    End If

                    System.GC.Collect()
                    System.GC.WaitForPendingFinalizers()
                End If

                ' TODO: nicht verwaltete Ressourcen (nicht verwaltete Objekte) freigeben und Finalize() weiter unten überschreiben.
                ' TODO: grosse Felder auf Null setzen.
            End If
            disposedValue = True
        End Sub



        ' TODO: Finalize() nur überschreiben, wenn Dispose(disposing As Boolean) weiter oben Code zur Bereinigung nicht verwalteter Ressourcen enthält.
        'Protected Overrides Sub Finalize()
        '    ' Ändern Sie diesen Code nicht. Fügen Sie Bereinigungscode in Dispose(disposing As Boolean) weiter oben ein.
        '    Dispose(False)
        '    MyBase.Finalize()
        'End Sub


        ' Dieser Code wird von Visual Basic hinzugefügt, um das Dispose-Muster richtig zu implementieren.
        Public Sub Dispose() Implements System.IDisposable.Dispose
            ' Ändern Sie diesen Code nicht. Fügen Sie Bereinigungscode in Dispose(disposing As Boolean) weiter oben ein.
            Dispose(True)
            ' TODO: Auskommentierung der folgenden Zeile aufheben, wenn Finalize() oben überschrieben wird.
            ' GC.SuppressFinalize(Me)
        End Sub



        Public Sub New()
            Me.New(New ParameterCollection)
        End Sub


        Public Sub New(ByVal allParameters As ParameterCollection)
            Me.Parameters = allParameters
        End Sub

        Public MustOverride Function EvaluateExpression() As Object

        ' Public Parameters As ParameterCollection
        ' Public MustOverride Sub SetCompulsoryParameter()


    End Class ' AbstractEvaluator


    ' example
    Public Class ConcreteEvaluator
        Inherits AbstractEvaluator

        Class ConcreteCode

            Public Function Tomorrow() As System.DateTime
                Return System.DateTime.Now.AddDays(1)
            End Function

        End Class

        Friend Code As New ConcreteCode


        Public Sub New()
            Me.New(New ParameterCollection)
        End Sub


        Public Sub New(ByVal allParameters As ParameterCollection)
            MyBase.New(allParameters)
            'code
        End Sub


        'Public Overrides Sub SetCompulsoryParameter()
        'End Sub

        Public Overrides Function EvaluateExpression() As Object
            'Dim expression As String = "System.DateTime.Now.AddDays(1+2+3).ToString(""dd.MM.yyyy"")" ' string expression would come from report, compiled with roslyn
            'Return " + expression + @"

            ' Parameters!in_standort.Value

            Return Code.Tomorrow().ToString("dd.MM.yyyy")
        End Function


    End Class


End Namespace

```

[![Haltemodus][1]][1]
[![Console][2]][2]

  [1]: https://i.stack.imgur.com/5TAkb.png
  [2]: https://i.stack.imgur.com/iC13g.png

**Comment by jkotas** [comment]
This is likely duplicate of https://github.com/dotnet/runtime/pull/32311

@ststeiger Could you please install latest 3.1 update from https://dotnet.microsoft.com/download/dotnet-core/3.1 and check whether it fixes the problem?

**Comment by tommcdon** [comment]
@ststeiger We are hoping that the issue is resolved with the latest 3.1 update.  Can you confirm that the issue resolved?

**Comment by tommcdon** [comment]
Closing due to no response
